pkg/cosign/tuf.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2024 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cosign
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/sigstore/sigstore-go/pkg/tuf"
+)
+
+const (
+	tufMirrorEnvVar   = "TUF_MIRROR"
+	tufRootDirEnvVar  = "TUF_ROOT"
+	tufRootJSONEnvVar = "TUF_ROOT_JSON"
+)
+
+type tufStatus int
+
+const (
+	active tufStatus = iota
+	inactive
+)
+
+func addKeyFromFile(publicKeys *TrustedTransparencyLogPubKeys, name, description string) (*TrustedTransparencyLogPubKeys, error) {
+	raw, err := os.ReadFile(name)
+	if err != nil {
+		return nil, fmt.Errorf("error reading alternate %s file: %w", description, err)
+	}
+	if err := publicKeys.AddTransparencyLogPubKey(raw, active); err != nil {
+		return nil, fmt.Errorf("error adding %s: %w", description, err)
+	}
+	return publicKeys, nil
+}
+
+func setTUFOpts() (*tuf.Options, error) {
+	opts := tuf.DefaultOptions()
+	if tufMirror := os.Getenv(tufMirrorEnvVar); tufMirror != "" { //nolint:forbidigo
+		opts.RepositoryBaseURL = tufMirror
+	}
+	if tufRootJSON := os.Getenv(tufRootJSONEnvVar); tufRootJSON != "" { //nolint:forbidigo
+		rootJSONBytes, err := os.ReadFile(tufRootJSON)
+		if err != nil {
+			return nil, fmt.Errorf("error reading TUF root.json: %w", err)
+		}
+		opts.Root = rootJSONBytes
+	}
+	if tufCacheDir := os.Getenv(tufRootDirEnvVar); tufCacheDir != "" { //nolint:forbidigo
+		opts.CachePath = tufCacheDir
+	}
+	return opts, nil
+}
+
+func addKeyFromTUF(publicKeys *TrustedTransparencyLogPubKeys, opts *tuf.Options, name, description string) (*TrustedTransparencyLogPubKeys, error) {
+	tufClient, err := tuf.New(opts)
+	if err != nil {
+		return nil, fmt.Errorf("error creating TUF client: %w", err)
+	}
+	pubKeyBytes, err := tufClient.GetTarget(name)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching %s: %w", description, err)
+	}
+	if err := publicKeys.AddTransparencyLogPubKey(pubKeyBytes, active); err != nil {
+		return nil, fmt.Errorf("error adding %s: %w", description, err)
+	}
+	return publicKeys, nil
+}
+
+func checkValidityPeriod(start, end time.Time) tufStatus {
+	now := time.Now()
+	if now.Before(start) {
+		return inactive
+	}
+	if now.After(end) {
+		return inactive
+	}
+	return active
+}
```

Comment:
The validity period in the TrustedRoot should be compared to the timestamp provided by the timestamping service or transparency log. I understand this is not something that is known at the time that you are assembling the CheckOpts, and that makes this a hard problem, but I'm not sure we want to enforce this in this version of the code. I believe we still want to be able to verify a signature produced prior to a timestamping service's validity end date, even if the current time is after that date.

Reply:
My goal here was to have something to substitute for the `Active`/`Expired` [status](https://github.com/sigstore/sigstore/blob/745a80351bcaac1f814672fa14460f1f20ef0ff8/pkg/tuf/status_type.go#L26-L27) from TUF v1 that does not seem to have an equivalent in TUF v2. Open to discarding this entirely or finding an alternative.

Reply:
For what it's worth, I don't think Cosign is effectively using Active/Expired now, as Cosign is only printing a message to stdout if you are verifying an entry with an "expired" key. This isn't really helpful to the user because there's nothing wrong with using an expired key, it's if the key is used to verify something outside of its validity window, which isn't currently implemented in Cosign.

If it'd be easier to just not deal with Active/Expired, I'd be supportive of that. We can either rely on sigstore-go down the line to do that check, or we can file an issue to implement that feature in Cosign's verification API when using the trusted root file. I'd lean towards the latter in line with https://github.com/sigstore/cosign/issues/3879.